src/control/cmd/daos/container.go

```diff
@@ -101,7 +101,7 @@ func (cmd *containerBaseCmd) openContainer() error {
 		cmd.log.Debugf("opening container: %s", cmd.contUUID)
 		cUUIDstr := C.CString(cmd.contUUID.String())
 		defer freeString(cUUIDstr)
-		rc = C.daos_cont_open(cmd.cPoolHandle, cUUIDstr,
+		rc = C.daos_cont_open2(cmd.cPoolHandle, cUUIDstr,
```

Comment:
in this case, it's probably better to call daos_cont_open with the uuid and not converting the uuid to a string?

Reply:
I tend to agree with that but didn't really want to touch too much go code

Reply:
Ah, missed this detail. So in this case you can just get the uuid with this:
```go
cUUID := uuidToC(cmd.contUUID)
```

That will return a uuid_t that you can use with whichever version of cont_open makes sense.

Reply:
It doesn't seem to work to pass a uuid_t to the function.

In my investigation, it appears that the macros work. However, when it looks at the "type" of the routine, it seems to ignore the macro for that test.

I tested with the following go code.   It won't compile because of this

jvolivie-desk1:[~/daos]$ go build -mod vendor main.go
# command-line-arguments
./main.go:27:25: cannot use uuidToC(myuuid) (type _Ctype_uuid_t) as type *_Ctype_char in argument to _Cfunc_quick_test

If I change the type of quick_test to take uuid_t, the next two lines complain.

Interestingly, if I remove the line above it compiles and quick_test2 gets invoked for both but the type checking seems to ignore macros.

`
//
// (C) Copyright 2021 Intel Corporation.
//
// SPDX-License-Identifier: BSD-2-Clause-Patent
//
package main

import (
       "github.com/google/uuid"
)

/*
#include "util.h"
*/
import "C"

func uuidToC(in uuid.UUID) (out C.uuid_t) {
     for i, v := range in {
        out[i] = C.uchar(v)
     }

     return
}

func main() {
    myuuid := uuid.New()
    C.quick_test(uuidToC(myuuid))
    C.quick_test(C.CString(myuuid.String()))
    C.quick_test(C.CString("Hello"))
}

`

With this header file

`//
// (C) Copyright 2021 Intel Corporation.
//
// SPDX-License-Identifier: BSD-2-Clause-Patent
//
package main

import (
       "github.com/google/uuid"
)

/*
#include "util.h"
*/
import "C"

func uuidToC(in uuid.UUID) (out C.uuid_t) {
     for i, v := range in {
        out[i] = C.uchar(v)
     }

     return
}

func main() {
    myuuid := uuid.New()
    C.quick_test(uuidToC(myuuid))
    C.quick_test(C.CString(myuuid.String()))
    C.quick_test(C.CString("Hello"))
}

jvolivie-desk1:[~/daos]$ cat util.h 
#include <uuid/uuid.h>
#include <stdio.h>
int quick_test(const char *);
int quick_test2(const char *);

#define d_is_uuid(var)                                                          \
        (__builtin_types_compatible_p(typeof(var), uuid_t) ||                   \
         __builtin_types_compatible_p(typeof(var), unsigned char *) ||          \
         __builtin_types_compatible_p(typeof(var), const unsigned char *) ||    \
         __builtin_types_compatible_p(typeof(var), const uuid_t))

#define quick_test(x)					\
	({								\
		int _ret;						\
		char _str[37];						\
		const char *__str = NULL;				\
		if (d_is_uuid(x)) {				\
			uuid_unparse((unsigned char *)(x), _str);	\
			__str = _str;					\
			printf(#x "is uuid\n");\
		} else {					\
			__str = (const char *)(x);			\
			printf(#x "is string\n");\
		}						\
		_ret = quick_test2((const char *)x);	\
		_ret;							\
	})
`